discovery/ionos/server.go

```diff
@@ -0,0 +1,170 @@
+
+package ionos
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	ionoscloud "github.com/ionos-cloud/sdk-go/v6"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	serverLabelPrefix = metaLabelPrefix + "server_"
+
+	serverAvailabilityZoneLabel = serverLabelPrefix + "availability_zone"
+	serverBootCDROMIDLabel      = serverLabelPrefix + "boot_cdrom_id"
+	serverBootImageIDLabel      = serverLabelPrefix + "boot_image_id"
+	serverBootVolumeIDLabel     = serverLabelPrefix + "boot_volume_id"
+	serverCPUFamilyLabel        = serverLabelPrefix + "cpu_family"
+	serverDatacenterIDLabel     = serverLabelPrefix + "datacenter_id"
+	serverIDLabel               = serverLabelPrefix + "id"
+	serverIPLabel               = serverLabelPrefix + "ip"
+	serverLANIDLabel            = serverLabelPrefix + "lan_id"
+	serverLifecycleLabel        = serverLabelPrefix + "lifecycle"
+	serverNameLabel             = serverLabelPrefix + "name"
+	serverNICIDLabel            = serverLabelPrefix + "nic_id"
+	serverPrimaryIPLabel        = serverLabelPrefix + "primary_ip"
+	serverPrimaryLANIDLabel     = serverLabelPrefix + "primary_lan_id"
+	serverPrimaryNICIDLabel     = serverLabelPrefix + "primary_nic_id"
+	serverStateLabel            = serverLabelPrefix + "state"
+	serverTypeLabel             = serverLabelPrefix + "type"
+)
+
+type serverDiscovery struct {
+	*refresh.Discovery
+	client       *ionoscloud.APIClient
+	port         int
+	datacenterID string
+}
+
+func newServerDiscovery(conf *SDConfig, logger log.Logger) (*serverDiscovery, error) {
+	d := &serverDiscovery{
+		port:         conf.Port,
+		datacenterID: conf.DatacenterID,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "ionos_sd")
+	if err != nil {
+		return nil, err
+	}
+
+	// username, password and token are set via http client config
+	cfg := ionoscloud.NewConfiguration("", "", "", conf.ionosEndpoint)
+	cfg.HTTPClient = &http.Client{
+		Transport: rt,
+		Timeout:   time.Duration(conf.RefreshInterval),
+	}
+	cfg.UserAgent = fmt.Sprintf("Prometheus/%s", version.Version)
+
+	d.client = ionoscloud.NewAPIClient(cfg)
+
+	return d, nil
+}
+
+func (d *serverDiscovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	api := d.client.ServersApi
+
+	servers, _, err := api.DatacentersServersGet(ctx, d.datacenterID).
+		Depth(3).
+		Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	var targets []model.LabelSet
+	for _, server := range *servers.Items {
+		labels := model.LabelSet{
+			serverAvailabilityZoneLabel: model.LabelValue(*server.Properties.AvailabilityZone),
+			serverCPUFamilyLabel:        model.LabelValue(*server.Properties.CpuFamily),
+			serverDatacenterIDLabel:     model.LabelValue(d.datacenterID),
```

Comment:
Could this come from the API instead? I can see how we could have different values, e.g. if it is case insensitive.

Reply:
The API doesn't provide a direct method of returning the datacenter ID from a server or list of servers. But the `servers` object has an `Id` attribute which returns `<datecenter_id>/servers` from which it should be possible to "compute" the datacenter ID. I thought about doing this, but wasn't sure if it would be worth it or not.

Reply:
if it is not exposed by the API I think we should expose the ID as is.

Reply:
I still think this

Reply:
I think I've misunderstood you. Your suggestion is to expose the ID returned by the servers object or not?

Reply:
The ID as is, yes

Reply:
Done.